File: solutions/day02/solution.py
import itertools
import numpy as np


def paper_required(dims):
    dims = map(int, dims.split('x'))
    sides = [np.prod(s) for s in itertools.combinations(dims, 2)]
    return sum([2 * s for s in sides]) + min(sides)


def ribbon_required(dims):
    dims = map(int, dims.split('x'))
    maxdim = max(dims)
    area = np.prod(dims)
    dims.remove(maxdim)
    perim = sum([2 * dim for dim in dims])
    return area + perim


with open('input.txt', 'r') as fh:
    input = fh.read().split()


print('part 1: {}'.format(sum(map(paper_required, input))))
print('part 1: {}'.format(sum(map(ribbon_required, input))))

File: solutions/day02/solution.scala
val input = scala.io.Source.fromFile("input.txt").getLines.map(_.split("x").map(_.toInt)).toList

def dimsToSides(dims: Array[Int]) = {
    // A shame - can't use combinations, because it returns the set of unique
    // combinations, which would potentially discard one or more sides
    // val sides = dims.combinations(2).toList
    val Array(l, w, h) = dims
    List(List(l, w), List(w, h), List(h, l))
}

def paperRequired(dims: Array[Int]) = {
    val sides = dimsToSides(dims)
    val side_areas = sides.map(_.product) 
    val smallest_side = side_areas.min 
    val total_area = side_areas.sum * 2
    total_area + smallest_side 
}

def ribbonRequired(dims: Array[Int]) = {
    val volume = dims.product 
    val sides = dimsToSides(dims)
    val perimeters = sides.map(_.sum * 2)
    volume + perimeters.min
}

println(s"part 1: ${input.map(box => paperRequired(box)).sum}")
println(s"part 2: ${input.map(box => ribbonRequired(box)).sum}")
